### PR TITLE
feat(cdn): add upload artifact handler

### PR DESCRIPTION
**Pull request by sguiheux**
- New Buffer that use local FS
- Upload artifact Handler

@ovh/cds

### CODE REVIEW

File: engine/cdn/cdn_sync.go [hunk 1]
```diff
@@ -38,39 +38,39 @@ func (s *Service) startCDSSync(ctx context.Context) error {
 func (s *Service) SyncBuffer(ctx context.Context) {
```
Comment (fsamin) [inline engine/cdn/cdn_sync.go:38]: probably this function should also be part of RunningStorageUnits

File: engine/cdn/item/gorp_model.go [hunk 2]
```diff
@@ -12,10 +14,32 @@ func InitDBMapping(m *gorpmapper.Mapper) {
...
+		var apiRef sdk.CDNArtifactAPIRef
+		if err := json.Unmarshal(c.APIRefDB, &apiRef); err != nil {
+			return item, sdk.WithStack(err)
+		}
+		item.APIRef = &apiRef
+	}
+	return item, nil
 }
 
 func toItemDB(c sdk.CDNItem) *cdnItemDB {
```
Comment (fsamin) [inline engine/cdn/item/gorp_model.go:39]: don't return a pointer

File: engine/cdn/storage/cds/cds.go [hunk 3]
```diff
@@ -56,14 +56,22 @@ func (c *CDS) NewWriter(_ context.Context, _ sdk.CDNItemUnit) (io.WriteCloser, e
 func (c *CDS) NewReader(ctx context.Context, i sdk.CDNItemUnit) (io.ReadCloser, error) {
 	switch i.Item.Type {
 	case sdk.CDNTypeItemStepLog:
-		bs, err := c.client.WorkflowNodeRunJobStepLog(ctx, i.Item.APIRef.ProjectKey, i.Item.APIRef.WorkflowName, i.Item.APIRef.NodeRunID, i.Item.APIRef.NodeRunJobID, i.Item.APIRef.StepOrder)
+		logApiRef, is := i.Item.APIRef.(*sdk.CDNLogAPIRef)
```
Comment (fsamin) [inline engine/cdn/storage/cds/cds.go:59]: put this statement in a nicer method of `sdk.CDNItem`

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by yesnault — APPROVED** [review]

File: engine/cdn/cdn_file.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	it.Hash = sha512S
+	it.MD5 = md5S
+	it.Size = sizeWriter.Size
+	it.Status = sdk.CDNStatusItemCompleted
+
+	// Insert Item and ItemUnit in database
+	tx, err := s.mustDBWithCtx(ctx).Begin()
+	if err != nil {
+		return sdk.WithStack(err)
+	}
```
Comment (richardlt) [inline engine/cdn/cdn_file.go:89]: defer tx rollback

File: engine/cdn/item_upload.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		}
+
+		// Verify Signature
+		if err := jws.Verify(workerData.PrivateKey, signatureString, &signature); err != nil {
+			return sdk.WrapError(err, "worker key: %d", len(workerData.PrivateKey))
+		}
+
+		reader, err := r.MultipartReader()
+		if err != nil {
+			return err
```
Comment (richardlt) [inline engine/cdn/item_upload.go:38]:
```suggestion
			return sdk.WithStack(err)
```

File: engine/cdn/item_upload.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		}
+		for {
+			part, err := reader.NextPart()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
```
Comment (richardlt) [inline engine/cdn/item_upload.go:46]:
```suggestion
				return sdk.WithStack(err)
```

File: engine/cdn/storage/storageunit.go [hunk 7]
```diff
@@ -87,6 +94,20 @@ func Init(ctx context.Context, m *gorpmapper.Mapper, store cache.Store, db *gorp
 				return nil, err
 			}
 			bufferUnit = bd
+		case bu.Local != nil:
+			d := GetDriver("local-buffer")
+			if d == nil {
+				return nil, fmt.Errorf("local driver is not available")
```
Comment (richardlt) [inline engine/cdn/storage/storageunit.go:100]:
```suggestion
				return nil, sdk.WithStack(fmt.Errorf("local driver is not available"))
```

File: engine/cdn/storage/storageunit.go [hunk 8]
```diff
@@ -87,6 +94,20 @@ func Init(ctx context.Context, m *gorpmapper.Mapper, store cache.Store, db *gorp
...
+			}
+			bd, is := d.(BufferUnit)
+			if !is {
+				return nil, fmt.Errorf("local driver is not a buffer unit driver")
```
Comment (richardlt) [inline engine/cdn/storage/storageunit.go:104]:
```suggestion
				return nil, sdk.WithStack(fmt.Errorf("local driver is not a buffer unit driver"))
```

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15279.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by fsamin — APPROVED** [review]

**Pull request merged**